File: stepik-union/src/main/java/org/stepik/plugin/projectView/StepikDirectoryNode.java
package org.stepik.plugin.projectView;

import com.intellij.ide.projectView.PresentationData;
import com.intellij.ide.projectView.ViewSettings;
import com.intellij.ide.projectView.impl.nodes.PsiDirectoryNode;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiDirectory;
import com.jetbrains.tmp.learning.StepikProjectManager;
import com.jetbrains.tmp.learning.StudyUtils;
import com.jetbrains.tmp.learning.courseFormat.StudyNode;
import org.jetbrains.annotations.NotNull;
import org.stepik.plugin.utils.PresentationDataUtils;

import static org.stepik.plugin.utils.ProjectPsiFilesUtils.getRelativePath;

/**
 * @author meanmail
 */
class StepikDirectoryNode extends PsiDirectoryNode {

    StepikDirectoryNode(
            Project project,
            PsiDirectory value,
            ViewSettings viewSettings) {
        super(project, value, viewSettings);
    }

    @Override
    protected void updateImpl(@NotNull PresentationData data) {
        PresentationDataUtils.updatePresentationData(data, getValue());
    }

    @Override
    public int getTypeSortWeight(boolean sortByType) {
        StudyNode node = StudyUtils.getStudyNode(myProject, getValue().getVirtualFile());

        String path = getRelativePath(getValue());

        if (node != null && node.getPath().equals(path)) {
            return node.getPosition();
        }

        return Integer.MAX_VALUE;
    }

    @Override
    public boolean canNavigate() {
        return true;
    }

    @Override
    public boolean canNavigateToSource() {
        return true;
    }

    @Override
    protected boolean hasProblemFileBeneath() {
        return false;
    }

    @Override
    public String getNavigateActionText(boolean focusEditor) {
        return null;
    }

    @Override
    public void navigate(boolean requestFocus) {
        VirtualFile virtualFile = getVirtualFile();
        StudyNode studyNode;
        if (virtualFile != null) {
            studyNode = StudyUtils.getStudyNode(myProject, virtualFile);
            StepikProjectManager.setSelected(myProject, studyNode);
        }
    }
}

File: stepik-union/src/main/kotlin/org/stepik/core/projectView/StepikDirectoryNode.kt
package org.stepik.core.projectView

import com.intellij.ide.projectView.PresentationData
import com.intellij.ide.projectView.ViewSettings
import com.intellij.ide.projectView.impl.nodes.PsiDirectoryNode
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiDirectory
import org.stepik.core.StepikProjectManager
import org.stepik.core.StudyUtils
import org.stepik.plugin.utils.PresentationDataUtils
import org.stepik.plugin.utils.ProjectPsiFilesUtils.getRelativePath


internal class StepikDirectoryNode(
        project: Project,
        value: PsiDirectory,
        viewSettings: ViewSettings) : PsiDirectoryNode(project, value, viewSettings) {

    override fun updateImpl(data: PresentationData) {
        PresentationDataUtils.updatePresentationData(data, value)
    }

    override fun getTypeSortWeight(sortByType: Boolean): Int {
        val node = StudyUtils.getStudyNode(myProject, value.virtualFile)

        return if (node?.path == getRelativePath(value)) {
            node.position
        } else Integer.MAX_VALUE
    }

    override fun canNavigate() = true

    override fun canNavigateToSource() = true

    override fun hasProblemFileBeneath() = false

    override fun getNavigateActionText(focusEditor: Boolean): String? = null

    override fun navigate(requestFocus: Boolean) {
        val virtualFile = virtualFile
        if (virtualFile != null) {
            val studyNode = StudyUtils.getStudyNode(myProject, virtualFile)
            StepikProjectManager.setSelected(myProject, studyNode)
        }
    }
}
